programs/btc_spv_api/src/spv_processor.rs

```diff
@@ -0,0 +1,56 @@
+//! Bitcoin SPV proof verifier program
+//! Receive merkle proofs and block headers, validate transaction
+use crate::exchange_state::*;
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::instruction::InstructionError;
+
+
+pub struct SpvProcessor {}
+
+impl SpvProcessor {
+
+    pub fn validateHeaderChain(headers: &HeaderChain, proof_req: &ProofRequest) -> Result<(), InstructionError> {
+        let ln            = *headers.len();
+        let confirmations = *proof_req.confirmations;
+        // check that the headerchain is the right length
+        if ln != (2 + confirmations) {
+            error!("Invalid length for Header Chain");
+            Err(InstructionError::InvalidArgument)? //is this ? necessary?
+        }
+        // check that the headerchain is in order and contiguous
+        for bh in 1..ln {
+            let prevHash  = *headers[bh-1].digest;
+            if *headers[bh].parent != prevHash {
+                error!("Invalid Header Chain sequence");
+                Err(InstructionError::InvalidArgument)?
+            }
+        }
+        //not done yet, needs difficulty check
+        Ok(())
+    }
+
+
+
+}
+pub fn process_instruction(
+    _program_id: &Pubkey,
+    keyed_accounts: &mut [KeyedAccount],
+    data: &[u8],
+) -> Result<(), InstructionError> {
+    solana_logger::setup();
+
+    let command = bincode::deserialize::<SpvInstruction>(data).map_err(|err| {
```

Comment:
how big can one of these be?

Reply:
proofsubmit is the biggest instruction by far, it has about 5-12 80 byte blockheaders to establish the context and confirmations as well as the actual merkle proof which is a vec of 32 byte hashes. 

The merkle proof length can vary based on the number of tx's per block (T) but its length should be log(T)/log(2) rounded up. So right now at 2.4k tx per block it would be 12 hashes long aka 384 bytes total. The max possible btc transactions per block is 16200 which would be a 14 hash long proof aka 448 bytes

so if the average proofrequest uses 7 confirmations and avg tx per block goes to 5k ish consistently (arbitrary realisticish numbers) then the size is roughly
(9 block headers * 80 )+ (13 merkle hashes * 32)  = 1.14KB

Reply:
solana maximum transaction size is right about that size...

Reply:
Ok, there's not much we can do to reduce the size since block headers and merkle hashes have a specific size but the other option would be limiting number of confirmations. We've also discussed having a trusted signer that submits blockheaders and having the verification program maintain them, but that's probably not a good long-term approach

Reply:
current going solution to this issue is to move the block headers to a separate account type that stores them in its data and can be referred to by requests or submitted proofs. This avoids the overhead of having to submit them for every proof while at the same offering a choice of header source on both sides of the transaction. This means that a client can submit a request with a reference to a specific header store of their preference and any proof submitter can fill their request, with the fees awarded being split between the headerStore Account and the Prover's account.
 It can also be used to set up a trusted header submitter system as previously discussed, but makes that purely opt in and provides a simple and frictionless way of transitioning away from that model as soon as market forces dictate

Reply:
@pathoam you can also feed them in chunks.  @rob-solana we can have a generic "data upload" program for feeding data onto the chain.

Reply:
ok good to know, that will be useful for setting up the headerchain accounts